extra/mariadbfrm/mariadb_frm.cc

```diff
@@ -0,0 +1,531 @@
+    my_free(buffer);
+    mysql_file_close(file, MYF(0));
+    return NULL;
+  }
+  
+  mysql_file_close(file, MYF(0));
+  return buffer;
+}
+
+/**
+  Extract database and table name from FRM file path
+*/
+static bool extract_db_table_names(const char *frm_path, 
+                                   LEX_CSTRING *db_name, 
+                                   LEX_CSTRING *table_name)
+{
+  char *path_copy, *db_start, *table_start, *frm_ext;
+  
+  if (!((path_copy= my_strdup(PSI_NOT_INSTRUMENTED, frm_path, MYF(MY_WME)))))
+    return true;
+  
+  if (!((frm_ext= strstr(path_copy, ".frm"))))
+  {
+    my_free(path_copy);
+    return true;
+  }
+  *frm_ext= '\0';
+  
+  table_start= strrchr(path_copy, '/');
+  if (!table_start)
+    table_start= strrchr(path_copy, '\\');
+  
+  if (!table_start)
+  {
+    table_name->str= my_strdup(PSI_NOT_INSTRUMENTED, path_copy, MYF(MY_WME));
+    table_name->length= strlen(table_name->str);
+    db_name->str= my_strdup(PSI_NOT_INSTRUMENTED, "test", MYF(MY_WME));
+    db_name->length= 4;
+    my_free(path_copy);
+    return false;
+  }
+  
+  *table_start= '\0';
+  table_start++;
+  
+  db_start= strrchr(path_copy, '/');
+  if (!db_start)
+    db_start= strrchr(path_copy, '\\');
+  
+  if (!db_start)
+    db_start= path_copy;
+  else
+    db_start++;
+  
+  table_name->str= my_strdup(PSI_NOT_INSTRUMENTED, table_start, MYF(MY_WME));
+  table_name->length= strlen(table_name->str);
+  db_name->str= my_strdup(PSI_NOT_INSTRUMENTED, db_start, MYF(MY_WME));
+  db_name->length= strlen(db_name->str);
+  
+  my_free(path_copy);
+  return false;
+}
+
+/**
+  Parse FRM file and create TABLE_SHARE and TABLE structures
+*/
+static bool parse_frm_file(THD *fake_thd, const char *frm_path)
+{
+  printf("DEBUG: Entering parse_frm_file\n");
+  fflush(stdout);
+  
+  TABLE_LIST table_list;
+  uchar *frm_data= NULL;
+  size_t frm_length= 0;
+  TABLE_SHARE *share= NULL;
+  TABLE *table= NULL;
+  LEX_CSTRING db_name, table_name;
+  bool error= true;
+  
+  memset(&table_list, 0, sizeof(table_list));
+  printf("DEBUG: table_list initialized\n");
+  fflush(stdout);
+
+  printf("DEBUG: About to read FRM file: %s\n", frm_path);
+  fflush(stdout);
+  
+  frm_data= read_frm_file(frm_path, &frm_length);
+  if (!frm_data)
+    goto cleanup;
+  
+  printf("DEBUG: FRM file read successfully, size: %zu bytes\n", frm_length);
+  fflush(stdout);
+
+  printf("DEBUG: Extracting database and table names\n");
+  fflush(stdout);
+  
+  if (extract_db_table_names(frm_path, &db_name, &table_name))
+  {
+    fprintf(stderr, "Error: Cannot extract database and table names from path\n");
+    goto cleanup;
+  }
+  
+  printf("DEBUG: Names extracted - db: %.*s, table: %.*s\n", 
+         (int)db_name.length, db_name.str, (int)table_name.length, table_name.str);
+  fflush(stdout);
+
+  printf("DEBUG: Allocating TABLE_SHARE manually\n");
+  fflush(stdout);
+  
+  share = (TABLE_SHARE*)my_malloc(PSI_NOT_INSTRUMENTED, sizeof(TABLE_SHARE), MYF(MY_ZEROFILL));
```

Comment:
No need. TABLE_SHARE is small enough to allocate it on stack. You'll get rid of an extra my_free as a bonus